### ISSUE TITLE
Option for recoverable NMI/exception

### ISSUE THREAD

**Issue by chuanhua** [riscv/riscv-isa-manual#261]
Although in the current privileged spec, NMI is used for hardware error conditions and is not recoverable, there are situations where NMI is used upon emergent dangerous system conditions, e.g., system voltage drop, to correct the system emergency immediately without being blocked and then resume to normal operation. 

To make NMI recoverable, I would like to propose a new layer of CSRs to save the states corrupted by NMI as follows.

New CSRS: mstack, mstack_epc, mstack_cause

When NMI happens, a hart enters “NMI” mode, and

- mstack.mpp <= mstatus.mpp <= interrupted-mode
- mstack.mpie <= mstatus.mpie <= mstatus.mie
- mstack_epc <= mepc <= interrupted-pc
- mstack_cause <= mcause <= interrupted-cause
- mstatus.mie <= 0; PC <= NMI handler address

When executing “mret” in “NMI” mode,

- mstack.mpp => mstatus.mpp => current-mode
- mstack.mpie => mstatus.mpie => mstatus.mie
- mstack_epc => mepc => PC
- mstack_cause => mcause

If recoverable NMI is not needed, with proper extension, this new layer of CSRs can be used to recover from an exception happened before an interrupt handler enables the global interrupt enable bit for nested interrupt or from a double exceptions (i.e., an exception happened in an exception handler). This time, the new layer of mstack CSRs are written upon exception entry, and are read upon executing mret instruction on exception exit respectively.

This is a shared scheme for both NMI and exception. A more expensive solution is to have one set of stack CSRs for NMI and another set of stack CSRs for exception.

**Comment by aswaterman** [comment]
Agreed we need to specify how this works, including the possibility of a resumable-NMI extension.

Not going to try to push this through prior to ratification.

**Comment by aswaterman** [comment]
We've handled this in the Smrnmi extension.
